### PR TITLE
Fix type assertion error in deployment controller DeleteFunc

### LINKED ISSUE #22208: Type assertion error in DeleteFunc of deployment controller
Found this when debugging #22122 

https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce/12453/artifacts/?debugUI=CLOUD
```
E0228 19:06:47.254179       5 runtime.go:58] Recovered from panic: &runtime.TypeAssertionError{interfaceString:"", concreteString:"cache.DeletedFinalStateUnknown", assertedString:"*extensions.Deployment", missingMethod:""} (interface conversion: interface is cache.DeletedFinalStateUnknown, not *extensions.Deployment)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:52
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:40
/usr/src/go/src/runtime/asm_amd64.s:402
/usr/src/go/src/runtime/panic.go:387
/usr/src/go/src/runtime/iface.go:225
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/deployment/deployment_controller.go:130
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:178
<autogenerated>:25
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:248
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:122
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:97
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:66
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:67
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:47
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:97
/usr/src/go/src/runtime/asm_amd64.s:2232
E0228 19:06:47.255209       5 reflector.go:333] /usr/src/go/src/runtime/asm_amd64.s:2232: expected type *api.PersistentVolumeClaim, but watch event object had type *api.Secret
```

@kubernetes/sig-config @bgrant0607 @soltysh @kargakis

### PR DESCRIPTION
**Pull request by janetkuo**
Fixes #22208

@kubernetes/sig-config @bgrant0607 @soltysh @kargakis

### CODE REVIEW

**Comment by soltysh** [comment]
LGTM

File: pkg/controller/deployment/deployment_controller.go [hunk 1]
```diff
@@ -127,7 +127,19 @@ func NewDeploymentController(client clientset.Interface, resyncPeriod controller
 			},
 			// This will enter the sync loop and no-op, because the deployment has been deleted from the store.
 			DeleteFunc: func(obj interface{}) {
-				d := obj.(*extensions.Deployment)
+				d, ok := obj.(*extensions.Deployment)
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: This probably should be moved into a function, as for the other types.

Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: Done

**Comment by janetkuo** [comment]
@bgrant0607 PTAL

**Comment by bgrant0607** [comment]
LGTM, thanks

**Pull request merged**